### PR TITLE
feat(nbstore): add sqlite implementation

### PR DESCRIPTION
**Pull request by forehalo**

### CODE REVIEW

**Comment by forehalo** [comment]
* **#8811** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8811?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8811?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8810** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8810?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8809** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8809?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8808** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8808?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#7639** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7639?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#8539](https://github.com/toeverything/AFFiNE/pull/8539) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8539?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `canary`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/toeverything?ref=merge-queue-instructions-in-stack-comment&prId=8811).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/common/nbstore/src/storage/storage.ts [hunk 1]
```diff
@@ -9,6 +9,22 @@
   id: string;
 }
 
+export function universalId(peer: string, type: SpaceType, id: string) {
+  return `@peer(${peer});@type(${type});@id(${id})`;
+}
+
+export function parseUniversalId(id: string) {
+  const match = id.match(/@peer\((.*)\);@type\((.*)\);@id\((.*)\)/);
```
[the remark this answers is not in the document]
Comment (forehalo) [reply packages/common/nbstore/src/storage/storage.ts]: resolved

**Review by Brooooooklyn — APPROVED** [review]

File: packages/frontend/apps/electron/src/helper/nbstore/doc.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  override async deleteDoc(docId: string) {
+    await this.db.deleteDoc(docId);
+  }
+
+  override async getDocTimestamps(after?: Date) {
+    const clocks = await this.db.getDocClocks(
+      after ? new Date(after) : undefined
+    );
```
Comment (darkskygit) [inline packages/frontend/apps/electron/src/helper/nbstore/doc.ts]: new Date again?

File: packages/common/nbstore/src/storage/storage.ts [hunk 3]
```diff
@@ -9,6 +9,77 @@ export interface StorageOptions {
...
+  for (const key of requiredKeys) {
+    if (!opts[key]) {
+      return false;
+    }
+  }
+
+  return opts.type === 'userspace' || opts.type === 'workspace';
+}
+
+export function parseUniversalId(id: string) {
```
Comment (darkskygit) [inline packages/common/nbstore/src/storage/storage.ts:32]: complex, why not split ';' then `/@(\w+)\((\w+)\)$/.match`

Comment (forehalo) [reply packages/common/nbstore/src/storage/storage.ts:32]: regex is more complex in parsing and even slower

Comment (darkskygit) [reply packages/common/nbstore/src/storage/storage.ts:32]: image

speed not bad in this case

Comment (forehalo) [reply packages/common/nbstore/src/storage/storage.ts:32]: but ';' will be treated as valid char in peer

Comment (darkskygit) [reply packages/common/nbstore/src/storage/storage.ts:32]: historical legacy issue?

Comment (forehalo) [reply packages/common/nbstore/src/storage/storage.ts:32]: nope, but peer is supposed to be any string

**Review by darkskygit — APPROVED** [review]

**Pull request merged**